libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1334,28 +1354,110 @@ string YulUtilFunctions::allocationFunction()
 	});
 }
 
-string YulUtilFunctions::allocateMemoryArrayFunction(ArrayType const& _type)
+string YulUtilFunctions::zeroMemoryArrayFunction(ArrayType const& _type)
 {
-	solUnimplementedAssert(!_type.isByteArray(), "");
+	if (_type.baseType()->hasSimpleZeroValueInMemory())
+		return zeroSimpleMemoryArrayFunction(_type);
+	return zeroComplexMemoryArrayFunction(_type);
+}
 
-	string functionName = "allocate_memory_array_" + _type.identifier();
+string YulUtilFunctions::zeroSimpleMemoryArrayFunction(ArrayType const& _type)
+{
+	solAssert(_type.baseType()->hasSimpleZeroValueInMemory(), "");
+
+	string functionName = "zero_simple_memory_array_" + _type.identifier();
 	return m_functionCollector.createFunction(functionName, [&]() {
 		return Whiskers(R"(
-			function <functionName>(length) -> memPtr {
-				memPtr := <alloc>(<allocSize>(length))
-				<?dynamic>
-				mstore(memPtr, length)
-				</dynamic>
+			function <functionName>(dataStart, dataSizeInBytes) {
+				calldatacopy(dataStart, calldatasize(), dataSizeInBytes)
 			}
 		)")
 		("functionName", functionName)
-		("alloc", allocationFunction())
-		("allocSize", arrayAllocationSizeFunction(_type))
-		("dynamic", _type.isDynamicallySized())
 		.render();
 	});
 }
 
+string YulUtilFunctions::zeroComplexMemoryArrayFunction(ArrayType const& _type)
+{
+	solAssert(!_type.baseType()->hasSimpleZeroValueInMemory(), "");
+
+	string functionName = "zero_complex_memory_array_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(dataStart, dataSizeInBytes) {
+				for {let i := 0} lt(i, dataSizeInBytes) { i := add(i, 32) } {
+					mstore(add(dataStart, i), <zeroValue>())
+				}
+			}
+		)")
+		("functionName", functionName)
+		("zeroValue", zeroValueFunction(*_type.baseType()))
+		.render();
+	});
+}
+
+string YulUtilFunctions::allocateAndInitializeMemoryArrayFunction(ArrayType const& _type)
+{
+	solUnimplementedAssert(!_type.isByteArray(), "");
+
+	string functionName = "allocate_and_zero_memory_array_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		auto const* base = _type.baseType();
+		solUnimplementedAssert(base->sizeOnStack() == 1, "Stacksize not yet implemented!");
+		return Whiskers(R"(
+				function <functionName>(length) -> memPtr {
+					let allocSize := <allocSize>(length)
+					memPtr := <alloc>(allocSize)
+					<?dynamic>
+					<zeroArrayFunction>(add(memPtr, 32), sub(allocSize, 32))
```

Comment:
I'm wondering if it would be cleaner to pass the length here instead.

Reply:
The called function needs to know where to start zeroing, or am I missing something?

Reply:
It's fine!